src/library/scala/collection/mutable/ArrayBuffer.scala

```diff
@@ -307,24 +306,29 @@ object ArrayBuffer extends StrictOptimizedSeqFactory[ArrayBuffer] {
 
   def empty[A]: ArrayBuffer[A] = new ArrayBuffer[A]()
 
-  import scala.runtime.PStatics.VM_MaxArraySize
+  @inline private def checkArrayLengthLimit(length: Int, currentLength: Int): Unit =
+    if (length > VM_MaxArraySize)
+      throw new Exception(s"Array of array-backed collection exceeds VM length limit of $VM_MaxArraySize. Requested length: $length; current length: $currentLength")
+    else if (length < 0)
+      throw new Exception(s"Overflow while resizing array of array-backed collection. Requested length: $length; current length: $currentLength; increase: ${length - currentLength}")
+
   /**
+   * The increased size for an array-backed collection.
+   *
    * @param arrayLen  the length of the backing array
    * @param targetLen the minimum length to resize up to
-   * @return -1 if no resizing is needed, the greater of targetLen and 2 * arrayLen if neither exceeds VM_MaxArraySize,
-   *         or VM_MaxArraySize otherwise.
+   * @return
+   *   - `-1` if no resizing is needed, else
+   *   - `VM_MaxArraySize` if `arrayLen` is too large to be doubled, else
+   *   - `max(targetLen, arrayLen * 2, , DefaultInitialSize)`.
+   *   - Throws an exception if `targetLen` exceeds `VM_MaxArraySize` or is negative (overflow).
    */
   private[mutable] def resizeUp(arrayLen: Int, targetLen: Int): Int = {
-    // Hybrid
     if (targetLen > 0 && targetLen <= arrayLen) -1
     else {
-      IterableOnce.checkArraySizeWithinVMLimit(targetLen) // safe because `targetSize <= Int.MaxValue`
-
-      math.min(
-        if (targetLen > (Int.MaxValue >> 1)) VM_MaxArraySize
-        else math.max(targetLen, math.max(arrayLen << 1, DefaultInitialSize)),
```

Comment:
ok, / and * instead of >> 1 and << 1.  Now I know.

Reply:
That's a standard thing. Shift has an aura of metal, but compilers know about powers of 2. Even I know how to double a number, at least up to ten fingers. So we prefer the notation that is easier to read.

Reply:
An exception is when you need to do correct division despite overflow.  `(a + b)/2` fails with signed math if `a` and `b` are sufficiently large (positive), but `(a + b) >>> 1` works correctly because signed and unsigned `a + b` have the same bit pattern, and `>>> 1` is a shift-as-if-unsigned.  Which necessarily brings it back into signed range.

This is WAY easier than the other logic you often need to do to avoid this edge case.